packages/framework/ai-collab/src/explicit-strategy/promptGeneration.ts

```diff
@@ -0,0 +1,301 @@
+	The tree is a JSON object with the following schema: ${promptFriendlySchema}
+	The state of the tree BEFORE changes was: ${originalDecoratedJson}.
+	The state of the tree AFTER changes is: ${decoratedTreeJson}.
+	The user requested that the following goal should be accomplished:
+	${userPrompt}
+	Was the goal accomplished?`;
+	return systemPrompt;
+}
+
+/**
+ * TBD
+ */
+export function getPromptFriendlyTreeSchema(jsonSchema: JsonTreeSchema): string {
+	let stringifiedSchema = "";
+	Object.entries(jsonSchema.$defs).forEach(([name, def]) => {
+		if (def.type !== "object" || def._treeNodeSchemaKind === NodeKind.Map) {
+			return;
+		}
+
+		let stringifiedEntry = `interface ${getFriendlySchemaName(name)} {`;
+
+		Object.entries(def.properties).forEach(([fieldName, fieldSchema]) => {
+			let typeString: string;
+			if (isJsonSchemaRef(fieldSchema)) {
+				const nextFieldName = fieldSchema.$ref;
+				const nextDef = getDef(jsonSchema.$defs, nextFieldName);
+				typeString = `${getTypeString(jsonSchema.$defs, [nextFieldName, nextDef])}`;
+			} else {
+				typeString = `${getAnyOfTypeString(jsonSchema.$defs, fieldSchema.anyOf, true)}`;
+			}
+			if (def.required && !def.required.includes(fieldName)) {
+				typeString = `${typeString} | undefined`;
+			}
+			stringifiedEntry += ` ${fieldName}: ${typeString};`;
+		});
+
+		stringifiedEntry += " }";
+
+		stringifiedSchema += (stringifiedSchema === "" ? "" : " ") + stringifiedEntry;
+	});
+	return stringifiedSchema;
+}
+
+function getTypeString(
+	defs: Record<string, JsonNodeSchema>,
+	[name, currentDef]: [string, JsonNodeSchema],
+): string {
+	const { _treeNodeSchemaKind } = currentDef;
+	if (_treeNodeSchemaKind === NodeKind.Leaf) {
+		return currentDef.type;
+	}
+	if (_treeNodeSchemaKind === NodeKind.Object) {
+		return getFriendlySchemaName(name);
+	}
+	if (_treeNodeSchemaKind === NodeKind.Array) {
+		const items = currentDef.items;
+		const innerType = isJsonSchemaRef(items)
+			? getTypeString(defs, [items.$ref, getDef(defs, items.$ref)])
+			: getAnyOfTypeString(defs, items.anyOf);
+		return `${innerType}[]`;
+	}
+	fail("Non-object, non-leaf, non-array schema type.");
+}
+
+function getAnyOfTypeString(
+	defs: Record<string, JsonNodeSchema>,
+	refList: JsonSchemaRef[],
+	topLevel = false,
+): string {
+	const typeNames: string[] = [];
+	refList.forEach((ref) => {
+		typeNames.push(getTypeString(defs, [ref.$ref, getDef(defs, ref.$ref)]));
+	});
+	const typeString = typeNames.join(" | ");
+	return topLevel ? typeString : `(${typeString})`;
+}
+
+function isJsonSchemaRef(field: JsonFieldSchema): field is JsonSchemaRef {
+	return (field as JsonSchemaRef).$ref !== undefined;
+}
+
+function getDef(defs: Record<string, JsonNodeSchema>, ref: string): JsonNodeSchema {
+	// strip the "#/$defs/" prefix
+	const strippedRef = ref.slice(8);
+	const nextDef = defs[strippedRef];
+	assert(nextDef !== undefined, "Ref not found.");
+	return nextDef;
+}
+
+/**
+ * TBD
```

Comment:
Converts a fully-qualified SharedTree schema name to a single-word name for use in textual TypeScript-style types.
TODO: Determine what to do with user-provided names that include periods (e.g. "Foo.Bar").
TODO: Should probably ensure name starts with an uppercase character.

Reply:
Added in commit b700ced170a403db24c887b25ce4917392948b3c